src/rdkafka_sasl_oauthbearer.c

```diff
@@ -0,0 +1,309 @@
+                "server-failure-message",
+        };
+        struct rd_kafka_sasl_oauthbearer_state *state = rktrans->rktrans_sasl.state;
+        rd_chariov_t out = RD_ZERO_INIT;
+        int r = -1;
+
+        rd_rkb_dbg(rktrans->rktrans_rkb, SECURITY, "SASLOAUTHBEARER",
+                   "SASL OAUTHBEARER client in state %s",
+                   state_names[state->state]);
+
+        switch (state->state)
+        {
+        case RD_KAFKA_SASL_OAUTHBEARER_STATE_SEND_CLIENT_FIRST_MESSAGE:
+                rd_dassert(!in); /* Not expecting any server-input */
+
+                rd_kafka_sasl_oauthbearer_build_client_first_message(rktrans, &out);
+                state->state = RD_KAFKA_SASL_OAUTHBEARER_STATE_RECEIVE_SERVER_FIRST_MESSAGE;
+                break;
+
+
+        case RD_KAFKA_SASL_OAUTHBEARER_STATE_RECEIVE_SERVER_FIRST_MESSAGE:
+                if (!in || in->ptr[0] == '\0') {
+                        /* Success */
+                        rd_rkb_dbg(rktrans->rktrans_rkb, SECURITY | RD_KAFKA_DBG_BROKER,
+                                "OAUTHBEARERAUTH",
+                                "SASL OAUTHBEARER authentication successful (principal=%s)",
+                                state->md_principal_name);
+                        rd_kafka_sasl_auth_done(rktrans);
+                        r = 0;
+                        break;
+                }
+
+                /* Failure; save error message for later */
+                state->server_error_msg.size = in->size;
+                state->server_error_msg.ptr  = rd_memdup(in->ptr,
+                                                state->server_error_msg.size);
+
+                /*
+                 * https://tools.ietf.org/html/rfc7628#section-3.1
+                 * kvsep          = %x01
+                 * client-resp    = (gs2-header kvsep *kvpair kvsep) / kvsep
+                 *
+                 * Send final kvsep (CTRL-A) character
+                 */
+                out.size = 1;
+                out.ptr = rd_malloc(out.size + 1);
+                rd_snprintf(out.ptr, out.size+1, "\x01");
+                state->state = RD_KAFKA_SASL_OAUTHBEARER_STATE_RECEIVE_SERVER_MESSAGE_AFTER_FAILURE;
+                r = 0; // will fail later in next state after sending current response
+                break;
+
+        case RD_KAFKA_SASL_OAUTHBEARER_STATE_RECEIVE_SERVER_MESSAGE_AFTER_FAILURE:
+                rd_dassert(!in); /* Not expecting any server-input */
+
+                /* Failure as previosuly communicated by server first message */
+                rd_snprintf(errstr, errstr_size, "SASL OAUTHBEARER authentication failed (principal=%s): %s",
+                        state->md_principal_name,
+                        state->server_error_msg.ptr);
+                rd_rkb_dbg(rktrans->rktrans_rkb, SECURITY | RD_KAFKA_DBG_BROKER,
+                        "OAUTHBEARERAUTH",
+                        "SASL OAUTHBEARER authentication failed (principal=%s): %s",
```

Comment:
since the errstr and the error log are identical you could reuse the errstr in the dbg log: `.., "%s", errstr);`

Reply:
👍